src/robot/running/arguments/typeconverters.py

```diff
@@ -389,61 +389,165 @@ class ListConverter(TypeConverter):
+        super().__init__(type_, languages=languages)
+        self.nested_types = getattr(type_, '__args__', ())
+
+    @classmethod
+    def handles(cls, type_):
+        base_type = getattr(type_, '__origin__', type_)
+        handled = (cls.type, cls.abc) if cls.abc else cls.type
+        if base_type is tuple:
             return False
-        return super().no_conversion_needed(value)
+        return isinstance(base_type, type) and issubclass(base_type, handled)
+
+    def no_conversion_needed(self, value):
+        return not self.nested_types \
+               and isinstance(value, self.used_type) \
+               and not isinstance(value, str)
 
     def _non_string_convert(self, value, explicit_type=True):
-        return list(value)
+        if isinstance(value, str):
+            converted_list = self._literal_eval(value, list)
+        elif is_list_like(value):
+            converted_list = list(value)
+        else:
+            raise ValueError
+        if self.nested_types:
+            if TypeConverter.converter_for(self.nested_types[0],
+                                           languages=self.languages):
+                for i, elem in enumerate(converted_list) :
+                    converter = TypeConverter.converter_for(self.nested_types[0],
+                                                            languages=self.languages)
+                    if converter:
+                        elem_info = f"List[{type_name(self.nested_types[0])}]"
+                        converted_list[i] = converter.convert(elem_info, elem,
+                                                              explicit_type)
+        return converted_list
 
-    def _convert(self, value, explicit_type=True):
-        return self._literal_eval(value, list)
+    _convert = _non_string_convert
 
 
 @TypeConverter.register
 class TupleConverter(TypeConverter):
     type = tuple
     type_name = 'tuple'
     value_types = (str, Sequence)
+    aliases = ('tuple',)
 
-    def _non_string_convert(self, value, explicit_type=True):
-        return tuple(value)
-
-    def _convert(self, value, explicit_type=True):
-        return self._literal_eval(value, tuple)
+    def __init__(self, type_, custom_converters=None, languages=None):
+        super().__init__(type_, languages=languages)
+        self.nested_types = getattr(type_, '__args__', ())
 
+    def _non_string_convert(self, value, explicit_type=True):
+        if isinstance(value, str):
+            converted_tuple = self._literal_eval(value, tuple)
+        elif is_list_like(value):
+            converted_tuple = tuple(value)
+        else:
+            raise ValueError
+        if self.nested_types:
+            tmp_list = list(converted_tuple)
+            for i, elem in enumerate(converted_tuple):
+                converter = TypeConverter.converter_for(self.nested_types[i],
+                                                        languages=self.languages)
```

Comment:
Yes, that is indeed a good idea. Note that for tuples it is not mandatory to list all types. Partial annotations by listing the first type(s) only is permitted.

Reply:
Is it documented somewhere that if you have e.g. `tuple[int, int, int]` type checkers should allow also `(0, 0, 0, 0)`? If not, I think we should be strict and require exactly the same amount of items as there are types listed. Otherwise there can be annoying runtime errors if you have e.g. `a, b, c = arg`. If you don't want such strict validation, you can use `list[int]` or `list[int|other|types]`.

Reply:
Never mind. I recalled reading this somewhere, but when looking it up again I came up empty. This is the official Python docu and it makes no such mention: https://docs.python.org/3/library/typing.html#typing.Tuple
Also, it is always possible to use the _Any_ type and still be complete.

So in short, I fully agree with your initial comment on this topic.

Reply:
The docs you referred mention `tuple[int, ...]` that we need to support. We could do it so that if the last type is `Ellipsis`, we allow tuples with more values that actual types. Not sure should we validate/convert those extra values or just leave them as-is. Related to that, we need to decide how to handle `tuple[int, str, ...]`.